yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -427,11 +474,25 @@ private[yarn] class YarnAllocator(
         containerSet.remove(containerId)
         if (containerSet.isEmpty) {
           allocatedHostToContainersMap.remove(host)
+          // If all the containers of this node are removed, remove this preferred location,
+          // since it means currently we don't need to this node locality.
+          preferredNodeLocationToUses.remove(host)
         } else {
           allocatedHostToContainersMap.update(host, containerSet)
         }
 
         allocatedContainerToHostMap.remove(containerId)
+
+        // If current there's no container allocated on this rack, remove it,
+        // since we don't need this rack currently.
+        val remainingRacks = allocatedHostToContainersMap.keys.map { h =>
+          RackResolver.resolve(conf, h).getNetworkLocation
```

Comment:
Probably not a huge deal, but isn't `RackResolver.resolve` a little expensive? IIRC it launches a subprocess that can do arbitrary things. Perhaps at some point it would be nice to cache this information.